src/telemetry/telemetryReporter.ts

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import Dap from '../dap/api';
+import { HighResolutionTime, calculateElapsedTime } from '../utils/performance';
+import { OpsReportBatcher, UnbatchedOpReport } from './opsReportBatch';
+import { EventEmitter } from '../common/events';
+
+export type TelemetryEntityProperties = object;
+export type OutcomeAndTime = { time: number, succesful: boolean };
+
+enum RequestOutcome {
+  Succesful,
+  Failed
+}
+
+export class DapRequestTelemetryReporter {
+  private readonly _rawTelemetryReporter: RawTelemetryReporterToDap;
+
+  public constructor(dap: Dap.Api) {
+    this._rawTelemetryReporter = new RawTelemetryReporterToDap(dap);
+  }
+
+  reportErrorWhileHandlingDapMessage(dapCommand: string, receivedTime: HighResolutionTime, error: unknown) {
+    this.reportWithElapsedTime(receivedTime, dapCommand, extractErrorDetails(error), RequestOutcome.Failed);
+  }
+
+  reportSuccesfullyHandledDapMessage(dapCommand: string, receivedTime: HighResolutionTime) {
+    this.reportWithElapsedTime(receivedTime, dapCommand, {}, RequestOutcome.Succesful);
+  }
+
+  private reportWithElapsedTime(receivedTime: [number, number], dapCommand: string, properties: TelemetryEntityProperties, outcome: RequestOutcome) {
+    const elapsedTime = calculateElapsedTime(receivedTime);
+    this._rawTelemetryReporter.report('dap/' + dapCommand, { ...properties, time: elapsedTime, succesful: outcome === RequestOutcome.Succesful });
+  }
+}
+
+export class CdpTelemetryReporter {
```

Comment:
This class seems almost identical to `DapRequestTelemetryReporter` except `reportWithElapsedTime` batches the events.

Can we consolidate the duplicated code between the two?

Reply:
Done